mrmustard/lab_dev/states/ket.py

```diff
@@ -273,6 +276,213 @@ def expectation(self, operator: CircuitComponent):
+        r"""
+        Applies the formal stellar decomposition.
+
+        Args:
+            core_modes: The set of modes defining core variables.
+
+        Returns:
+            S: The core state (`Ket`)
+            T: The Gaussian `Operation` performing the stellar decomposition.
+
+        Note:
+            This method pulls out the unitary ``U`` from the given state on the given modes, so that
+            the remaining state is a core state. Formally, we have
+            .. math::
+
+                \psi = (T\otimes\mathbb I) S_{\mathrm{core}}
+
+            where the operator :math:`T` acts on the given `core_modes` only.
+            Core states have favorable properties in the Fock representation
+            e.g., being sparse.
+
+        .. code-block::
+            >>> from mrmustard.lab_dev import Ket
+
+            >>> psi = Ket.random([0,1])
+            >>> core, t = psi.formal_stellar_decomposition([0])
+            >>> A_core, _, _ = core.ansatz.triple
+
+            >>> assert A_core[0,0] == 0
+        """
+        other_modes = [m for m in self.modes if m not in core_modes]
+        core_indices = self.wires[core_modes].indices
+        other_indices = self.wires[other_modes].indices
+        new_order = core_indices + other_indices
+
+        A, b, c = self.ansatz.reorder(new_order).triple
+
+        M = len(core_modes)
+
+        # we pick the blocks according to the naming chosen in the paper
+        Am = A[..., :M, :M]
+        R = A[..., M:, :M]
+        R_transpose = math.einsum("...ij->...ji", R)
+        An = A[..., M:, M:]
+        bm = b[..., :M]
+        bn = b[..., M:]
+
+        batch_shape = self.ansatz.batch_shape
+
+        As = math.block(
+            [
+                [
+                    math.zeros(
+                        batch_shape
+                        + (
+                            M,
+                            M,
+                        ),
+                        dtype=math.complex128,
+                    ),
+                    R_transpose,
+                ],
+                [R, An],
+            ]
+        )
+
+        bs = math.concat([math.zeros(batch_shape + (M,), dtype=math.complex128), bn], -1)
+        cs = c
+
+        inverse_order = np.argsort(new_order)
+        As = As[..., inverse_order, :]
+        As = As[..., :, inverse_order]
+        bs = bs[..., inverse_order]
+
+        s = Ket.from_bargmann(self.modes, (As, bs, cs))
+
+        if batch_shape != ():
+            Im = math.stack(
+                [math.eye(M, dtype=math.complex128)] * int(math.prod(batch_shape))
+            ).reshape(batch_shape + (M,) * 2)
+        else:
+            Im = math.eye(M, dtype=math.complex128)
+
+        At = math.block(
+            [
+                [Am, Im],
+                [
+                    Im,
+                    math.zeros(batch_shape + (M, M), dtype=math.complex128),
+                ],
+            ]
+        )
```

Comment:
```suggestion
        At = math.block([[Am, Im], [Im, zero]])
```
(using `zero` defined above)